### PR TITLE
Preparations for external transport of replicated events

### PR DESCRIPTION
**Pull request by johanandren**
* Ack protocol for the events being consumed
* A factory method to create replicated es entities without journal trailing through persistence query

### CODE REVIEW

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala [hunk 1]
```diff
@@ -416,6 +420,19 @@ private[akka] object Running {
...
+      val sideEffects = ackToOnPersisted match {
+        case None => Nil
+        case Some(ref) =>
+          SideEffect { (_: S) =>
+            // FIXME do we really have to create this every time?
+            val pid = ReplicationId(
+              setup.persistenceId.entityTypeHint,
+              setup.persistenceId.entityId,
+              event.originReplica).persistenceId
+            ref ! EventConsumed(pid, event.originSequenceNr)
```
Comment (patriknw) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala:432]: Do we need this ack message or can we use `Done`? If it's intended to be sent with `ask` the requester should already know the pid and seqNr?

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala [hunk 2]
```diff
@@ -343,6 +344,7 @@ private[akka] object Running {
               event.sequenceNumber,
               originReplicaId,
               expectedSequenceNumber)
+          event.replyTo.foreach(_ ! EventConsumed(event.persistenceId, event.sequenceNumber))
```
Comment (patriknw) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala:347]: wonder if we should also ack for the `originReplicaId == replication.replicaId` case, but maybe that isn't supposed to happen?

**Review by patriknw — COMMENTED** [review]
great to see that it fits in with the existing mechanism

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala (hunk 1 above)
Comment (johanandren) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala:432]: Yeah, might not be needed.

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/ReplicatedEventSourcing.scala [hunk 3]
```diff
@@ -122,4 +122,22 @@ object ReplicatedEventSourcing {
...
+  /**
+   * Initialize a replicated event sourced behavior.
+   *
+   * Events from each replica for the same entityId needs to be passed to it by an external stream.
+   * Care must be taken to handle events in any order as events can happen concurrently at different replicas.
+   *
+   * Using an replicated event sourced behavior means there is no longer the single writer guarantee.
+   *
+   * The journal plugin id for the entity itself can be configured using withJournalPluginId after creation.
+   */
```
Comment (pvlugter) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/ReplicatedEventSourcing.scala:134]:
```suggestion
  /**
   * Initialize a replicated event sourced behavior.
   *
   * Events from each replica for the same entityId need to be passed to it by an external stream.
   * Care must be taken to handle events in any order as events can happen concurrently at different replicas.
   *
   * Using a replicated event sourced behavior means there is no longer the single writer guarantee.
   *
   * The journal plugin id for the entity itself can be configured using withJournalPluginId after creation.
   */
```

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/ReplicatedEventSourcing.scala [hunk 4]
```diff
@@ -114,4 +114,20 @@ object ReplicatedEventSourcing {
...
+  /**
+   * Initialize a replicated event sourced behavior.
+   *
+   * Events from each replica for the same entityId needs to be passed to it by an external stream.
+   * Care must be taken to handle events in any order as events can happen concurrently at different replicas.
+   *
+   * Using an replicated event sourced behavior means there is no longer the single writer guarantee.
+   *
+   * The journal plugin id for the entity itself can be configured using withJournalPluginId after creation.
+   */
```
Comment (pvlugter) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/ReplicatedEventSourcing.scala:126]:
```suggestion
  /**
   * Initialize a replicated event sourced behavior.
   *
   * Events from each replica for the same entityId need to be passed to it by an external stream.
   * Care must be taken to handle events in any order as events can happen concurrently at different replicas.
   *
   * Using a replicated event sourced behavior means there is no longer the single writer guarantee.
   *
   * The journal plugin id for the entity itself can be configured using withJournalPluginId after creation.
   */
```

**Review by pvlugter — APPROVED** [review]
LGTM. 👍🏼 great that the existing mechanism can be used.

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala [hunk 5]
```diff
@@ -299,7 +302,9 @@ private[akka] object Running {
         case Some(replication) =>
           event.replicatedMetaData match {
             case None =>
-              setup.internalLogger.warn("Received published event for [{}] but with no replicated metadata, dropping")
+              setup.internalLogger.warn(
+                "Received published event for [{}] but with no replicated metadata, dropping",
+                event.persistenceId)
```
Comment (patriknw) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala:307]: 👍

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/ReplicatedEventSourcing.scala [hunk 6]
```diff
@@ -122,4 +122,22 @@ object ReplicatedEventSourcing {
...
+   * Initialize a replicated event sourced behavior.
+   *
+   * Events from each replica for the same entityId need to be passed to it by an external stream.
+   * Care must be taken to handle events in any order as events can happen concurrently at different replicas.
+   *
+   * Using a replicated event sourced behavior means there is no longer the single writer guarantee.
+   *
+   * The journal plugin id for the entity itself can be configured using withJournalPluginId after creation.
+   */
+  def externalReplication[Command, Event, State](replicationId: ReplicationId, allReplicas: JSet[ReplicaId])(
```
Comment (patriknw) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/ReplicatedEventSourcing.scala:135]: `@ApiMayChange` until we are more sure about the final api?

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/serialization/ReplicatedEventSourcingSerializer.scala [hunk 7]
```diff
@@ -153,7 +157,7 @@ import scala.collection.immutable.TreeMap
       .setPayload(wrappedSupport.payloadBuilder(impl.payload))
       .setTimestamp(impl.timestamp)
 
-    (impl.replicatedMetaData match {
+    impl.replicatedMetaData match {
       case None =>
         builder
```
Comment (patriknw) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/serialization/ReplicatedEventSourcingSerializer.scala:162]: this isn't needed now

**Review by patriknw — APPROVED** [review]
LGTM, after something minor

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/ReplicatedEventSourcing.scala (hunk 6 above)
Comment (johanandren) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/ReplicatedEventSourcing.scala:135]: Maybe I should just make it akka-internal for now, the way to actually publish events is anyway not public API.

Comment (patriknw) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/ReplicatedEventSourcing.scala:135]: sounds good `@InternalStableApi` (I know you don't like it, but public api come more drawbacks)

Comment (johanandren) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/ReplicatedEventSourcing.scala:135]: That also probably means we can skip the Java API, so less changes here. 👍

**Pull request merged**